### PR TITLE
Unbounded TextField width error

### LINKED ISSUE #9780: TextField needs to behave better when unbounded
Currently, it just throws this exception:

```
I/flutter (14542): ══╡ EXCEPTION CAUGHT BY RENDERING LIBRARY ╞═════════════════════════════════════════════════════════
I/flutter (14542): The following assertion was thrown during performLayout():
I/flutter (14542): BoxConstraints forces an infinite width.
I/flutter (14542): These invalid constraints were provided to RenderStack's layout() function by the following
I/flutter (14542): function, which probably computed the invalid constraints in question:
I/flutter (14542):   RenderConstrainedBox.performLayout (package:flutter/src/rendering/proxy_box.dart:260)
I/flutter (14542): The offending constraints were:
I/flutter (14542):   BoxConstraints(w=Infinity, 0.0<=h<=Infinity)
I/flutter (14542): 
I/flutter (14542): When the exception was thrown, this was the stack:
I/flutter (14542): #0      BoxConstraints.debugAssertIsValid.<anonymous closure>.throwError (package:flutter/src/rendering/box.dart:394)
I/flutter (14542): #1      BoxConstraints.debugAssertIsValid.<anonymous closure> (package:flutter/src/rendering/box.dart:435)
I/flutter (14542): #3      BoxConstraints.debugAssertIsValid (package:flutter/src/rendering/box.dart:389)
I/flutter (14542): #4      RenderObject.layout (package:flutter/src/rendering/object.dart:1789)
I/flutter (14542): #5      RenderConstrainedBox.performLayout (package:flutter/src/rendering/proxy_box.dart:260)
I/flutter (14542): #6      RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #7      RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #8      RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #9      RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #10     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #11     RenderFlex.performLayout (package:flutter/src/rendering/flex.dart:500)
I/flutter (14542): #12     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #13     RenderFlex.performLayout (package:flutter/src/rendering/flex.dart:500)
I/flutter (14542): #14     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #15     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #16     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #17     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #18     _RenderCustomClip.performLayout (package:flutter/src/rendering/proxy_box.dart:980)
I/flutter (14542): #19     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #20     MultiChildLayoutDelegate.layoutChild (package:flutter/src/rendering/custom_layout.dart:124)
I/flutter (14542): #21     _ScaffoldLayout.performLayout (package:flutter/src/material/scaffold.dart:85)
I/flutter (14542): #22     MultiChildLayoutDelegate._callPerformLayout (package:flutter/src/rendering/custom_layout.dart:194)
I/flutter (14542): #23     RenderCustomMultiChildLayoutBox.performLayout (package:flutter/src/rendering/custom_layout.dart:338)
I/flutter (14542): #24     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #25     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #26     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #27     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #28     _RenderCustomClip.performLayout (package:flutter/src/rendering/proxy_box.dart:980)
I/flutter (14542): #29     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #30     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #31     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #32     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #33     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #34     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #35     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #36     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #37     RenderOffstage.performLayout (package:flutter/src/rendering/proxy_box.dart:2483)
I/flutter (14542): #38     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #39     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #40     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #41     RenderStack.performLayout (package:flutter/src/rendering/stack.dart:433)
I/flutter (14542): #42     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #43     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #44     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #45     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #46     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #47     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #48     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #49     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #50     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #51     RenderBox&RenderObjectWithChildMixin&RenderProxyBoxMixin.performLayout (package:flutter/src/rendering/proxy_box.dart:100)
I/flutter (14542): #52     RenderObject.layout (package:flutter/src/rendering/object.dart:1872)
I/flutter (14542): #53     RenderView.performLayout (package:flutter/src/rendering/view.dart:120)
I/flutter (14542): #54     RenderObject._layoutWithoutResize (package:flutter/src/rendering/object.dart:1749)
I/flutter (14542): #55     PipelineOwner.flushLayout (package:flutter/src/rendering/object.dart:1050)
I/flutter (14542): #56     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:232)
I/flutter (14542): #57     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding&WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:361)
I/flutter (14542): #58     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:170)
I/flutter (14542): #59     BindingBase&SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:685)
I/flutter (14542): #60     BindingBase&SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:633)
I/flutter (14542): #61     BindingBase&SchedulerBinding&GestureBinding&ServicesBinding&RendererBinding.scheduleWarmUpFrame.<anonymous closure> (package:flutter/src/rendering/binding.dart:247)
I/flutter (14542): #63     _Timer._runTimers (dart:isolate-patch/timer_impl.dart:366)
I/flutter (14542): #64     _Timer._handleMessage (dart:isolate-patch/timer_impl.dart:394)
I/flutter (14542): #65     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:151)
I/flutter (14542): (elided 2 frames from class _AssertionError and package dart:async-patch)
I/flutter (14542): 
I/flutter (14542): The following RenderObject was being processed when the exception was fired:
I/flutter (14542):   RenderConstrainedBox#987687393 relayoutBoundary=up7 NEEDS-LAYOUT NEEDS-PAINT
I/flutter (14542):   creator: ConstrainedBox ← InputDecorator ← AnimatedBuilder ← Listener ← _GestureSemantics ←
I/flutter (14542):   RawGestureDetector ← GestureDetector ← TextField ← Row ← Column ← DefaultTextStyle ←
I/flutter (14542):   AnimatedDefaultTextStyle ← ⋯
I/flutter (14542):   parentData: <none> (can use size)
I/flutter (14542):   constraints: BoxConstraints(unconstrained)
I/flutter (14542):   size: MISSING
I/flutter (14542):   additionalConstraints: BoxConstraints(w=Infinity, 0.0<=h<=Infinity)
I/flutter (14542): This RenderObject had the following descendants (showing up to depth 5):
I/flutter (14542):   RenderStack#104464218 NEEDS-LAYOUT NEEDS-PAINT
I/flutter (14542):     RenderPadding#988890708 NEEDS-LAYOUT NEEDS-PAINT
I/flutter (14542):       RenderDecoratedBox#188031708 NEEDS-LAYOUT NEEDS-PAINT
I/flutter (14542):         RenderPadding#384224006 NEEDS-LAYOUT NEEDS-PAINT
I/flutter (14542):           RenderRepaintBoundary#914689605 NEEDS-LAYOUT NEEDS-PAINT
I/flutter (14542): ════════════════════════════════════════════════════════════════════════════════════════════════════
```

From this it's kinda hard to figure out what's wrong (especially for somebody new). I was able to fix the problem by wrapping the TextField in an Expanded widget.

/cc @Hixie

### PR DESCRIPTION
**Pull request by justinmc**
Closes: https://github.com/flutter/flutter/issues/9780

This improves the error message shown when a TextField ends up with an unbounded width.

### CODE REVIEW

**Comment by justinmc** [comment]
@HansMuller I included two options for where to put the assert here, but I plan to only include one once you help me figure out where the best place to put it is :grin:.   Let me know what you think.

File: packages/flutter/lib/src/material/input_decorator.dart [hunk 1]
```diff
@@ -834,6 +834,11 @@ class _RenderDecoration extends RenderBox {
   // This method applies layout to all of the renderers except the container.
   // For convenience, the container is laid out in performLayout().
   _RenderDecorationLayout _layout(BoxConstraints layoutConstraints) {
+    assert(
+      layoutConstraints.maxWidth < double.infinity,
+      'A text input cannot have an unbounded width. If you want it to take up the leftover space inside its parent, try wrapping it in a Flexible.',
```
Comment (HansMuller) [inline packages/flutter/lib/src/material/input_decorator.dart]: Maybe make the report a little more informative:

'An InputDecorator, which is typically created by a TextField, cannot have an unbounded width.\n'
'This happens when the parent widget does not provide a finite width constraint. '
'For example if the InputDecorator is contained by a Row, then its width must be '
'constrained. An Expanded widget or a SizedBox can be used to constrain the width '
'of the InputDecorator or the TextField that contains it.'

File: packages/flutter/lib/src/widgets/basic.dart [hunk 2]
```diff
@@ -3682,6 +3684,12 @@ class Flex extends MultiChildRenderObjectWidget {
        assert(crossAxisAlignment != null),
        assert(verticalDirection != null),
        assert(crossAxisAlignment != CrossAxisAlignment.baseline || textBaseline != null),
+       assert(
```
Comment (HansMuller) [inline packages/flutter/lib/src/widgets/basic.dart]: This doesn't seem like the place for this check. Maybe leave it out of this PR.

**Comment by justinmc** [comment]
@HansMuller Ready for a real review.  I eliminated the extra assert and updated the text on the remaining one.

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by gegobyte** [comment]
"An Expanded widget or a SizedBox can be used to constrain the width of the InputDecorator or the TextField that contains it"

Tried it with both SizedBox and Expanded and neither of them works.

**Comment by justinmc** [comment]
@gegobyte Can you open a new issue with a minimal app that reproduces the problem?  It's been awhile since I looked at this issue, but it should just be as easy as wrapping it: https://dartpad.dartlang.org/1115e38b012a9bc5c612ec3997b9367a

**Pull request merged**